### PR TITLE
380 show dialog and close all dialogs api

### LINKED ISSUE #380: Implement showDialog Api
**Is your feature request related to a problem? Please describe.**
Implement `showDialog` api.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by anserwaseem**
## Issue ticket number and link

Closes #380

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] I have added tests
- [x] I have added a changeset `pnpm changeset add`
- [x] I have added example usage in the kitchen sink app

### CODE REVIEW

File: packages/runtime/src/runtime/showDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  openModal?: (...args: any[]) => void;
+};
+
+export const showDialog = (props?: ShowDialogApiProps): void => {
+  const { action, openModal } = props ?? {};
+  if (!action || !openModal || !action?.widget) {
+    return;
+  }
+
+  const modalOptions = {
```
Comment (evshi) [inline packages/runtime/src/runtime/showDialog.tsx:24]: Can we make this consistent with the hook?

File: packages/runtime/src/runtime/showDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  let inlineWidget: EnsembleWidget | undefined;
+  let customWidget: React.FC | ReactElement | undefined;
+
+  if (isString(widget?.name)) {
+    customWidget = WidgetRegistry.find(widget?.name);
+  } else if (widgetInputs !== undefined) {
+    customWidget = WidgetRegistry.find(widgetKey || "");
+  } else {
+    inlineWidget = cloneDeep(unwrapWidget(action?.widget));
+  }
```
Comment (evshi) [inline packages/runtime/src/runtime/showDialog.tsx:53]: Don't go looking up the component function, just create a widget model and pass it to EnsembleRuntime.

File: packages/runtime/src/runtime/showDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  const WidgetFn = customWidget as React.FC<unknown>;
+
+  const content = (
+    <CustomScopeProvider
+      value={{
+        ...(widgetInputs as CustomScope),
+        ...(widget?.inputs as CustomScope),
+      }}
+    >
```
Comment (evshi) [inline packages/runtime/src/runtime/showDialog.tsx:63]: Custom widget already handles this if the model is set up correctly.

File: packages/runtime/src/runtime/customWidget.tsx [hunk 4]
```diff
@@ -39,20 +39,21 @@ const OnLoadAction: React.FC<
     context: Record<string, unknown>;
   }>
 > = ({ action, children, context }) => {
-  const onLoadAction = useEnsembleAction(action);
+  const onLoadActionRef = useRef(useEnsembleAction(action));
```
Comment (evshi) [inline packages/runtime/src/runtime/customWidget.tsx:42]: This is a sign you're breaking the render cycle in some way - why did you need this?

**Review by evshi — COMMENTED** [review]
Keep the implementation as close to the hook as possible - the more you diverge, it's a sign you're not headed in the right direction. Stick to the contract of the widget model and EnsembleRuntime should do most of the heavy lifting.

File: packages/runtime/src/runtime/customWidget.tsx (hunk 4 above)
Comment (anserwaseem) [reply packages/runtime/src/runtime/customWidget.tsx:42]: because custom widget's onLoad is firing twice :(

Comment (evshi) [reply packages/runtime/src/runtime/customWidget.tsx:42]: Like we discussed, if it's not obvious what the issue is, let's leave this implementation as is for now and do a follow up PR.

File: packages/runtime/src/runtime/showDialog.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    return;
+  }
+
+  const widget = action?.widget;
+  const widgetKey = findKey(widget, isObject) || "";
+  let inlineWidget: EnsembleWidget | undefined;
+
+  if ((get(widget, widgetKey) as Record<string, unknown>)?.inputs) {
+    set(widget, "name", widgetKey);
+    set(widget, "properties", get(widget, widgetKey));
```
Comment (evshi) [inline packages/runtime/src/runtime/showDialog.ts:27]: It's better to clone/create a new object here than modify the widget model inline. It's not clear to me why this requires special handling and can't be unwrapped with the util?

**Review by evshi — APPROVED** [review]
